docs/rfcs/0003-sql-datastore.md

```diff
@@ -0,0 +1,199 @@
+
+# Detailed design
+
+As mention is the motivation section, the root cause of this change is: we're going to add complex queries with those indexes to the datastore of PipeCD control-plane. The use-case of those queries requires:
+- easy way to pagination query result
+- easy way to ordering query result
+- easy way to add indexes for future adding queries
+
+Besides, to keep `simplicity on installing` characteristic of PipeCD, the chosen must be:
+- easy to be installed on cloud provider environment (GCP, AWS, etc)
+- can use the updated version easily
+
+We're using NoSQL for now due to its schemaless characteristic is good for the deployment process, so we focus on NoSQL support features on the chosen SQL database.
+
+The sample table creates commands for the `project` model and `application` model (other models than project model is as same as the application) as below.
+
+```sql
+# PostgreSQL
+CREATE TABLE projects (
+  id UUID PRIMARY KEY,
+  data JSONB NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at BIGINT NOT NULL,
+  updated_at BIGINT NOT NULL
+);
+CREATE TABLE applications (
+  id UUID PRIMARY KEY,
+  project_id UUID NOT NULL,
+  data JSONB NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at BIGINT NOT NULL,
+  updated_at BIGINT NOT NULL
+);
+
+# MySQL
+CREATE TABLE `projects` (
+  id BINARY(16) PRIMARY KEY,
+  data JSON NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at INT(11) NOT NULL,
+  updated_at INT(11) NOT NULL
+) ENGINE=InnoDB;
+
+CREATE TABLE `applications` (
+  id BINARY(16) PRIMARY KEY,
+  project_id BINARY(16) NOT NULL,
+  data JSON NOT NULL,
+  disabled BOOL NOT NULL,
+  created_at INT(11) NOT NULL,
+  updated_at INT(11) NOT NULL
+) ENGINE=InnoDB;
+```
+
+For indexing issue:
+- PostgreSQL indexing for the attributes of Jsonb data type is as follow: [Json indexing](https://www.postgresql.org/docs/current/datatype-json.html#JSON-INDEXING)
+- MySQL also support indexing for attributes of Json data type as follow: [Json indexing by secondary indexes](https://dev.mysql.com/doc/refman/8.0/en/create-table-secondary-indexes.html)
+
+(note: both Jsonb and Json point to the type of storing JSON data in those databases. The key difference is that Json data is stored as an exact copy of the JSON input text, whereas Jsonb stores data in a decomposed binary form. Jsonb seems a better choice since it supports indexing with significant performance increases)
+
+# Alternatives
+
+Currently, we consider between MySQL and PostgreSQL for those support for NoSQL features, PostgreSQL has a longer time in this field. Some considering factors:
+- ability to index a specific attribute of the JSON field
+- supported operators on specific JSON field
+- performance of query operation on indexed JSON field
+- able to keep the advantage of schemaless pattern
```

Comment:
Another question is that will we create multiple indexes or a single index combined multiple columns?